model/map-ldap/src/main/java/org/keycloak/models/map/storage/ldap/role/LdapRoleMapKeycloakTransaction.java

```diff
@@ -0,0 +1,412 @@
+        if (ldapObject != null) {
+            return new LdapMapRoleEntityFieldDelegate(new LdapRoleEntity(ldapObject, roleMapperConfig, this, clientId));
+        }
+        return null;
+    }
+
+    @Override
+    public Stream<MapRoleEntity> read(QueryParameters<RoleModel> queryParameters) {
+        LdapRoleModelCriteriaBuilder mcb = queryParameters.getModelCriteriaBuilder()
+                .flashToModelCriteriaBuilder(createLdapModelCriteriaBuilder());
+
+        Boolean isClientRole = mcb.isClientRole();
+        String clientId = mcb.getClientId();
+
+        LdapMapIdentityStore identityStore = new LdapMapIdentityStore(session, ldapMapConfig);
+
+        LdapMapQuery ldapQuery = getLdapQuery(isClientRole, clientId);
+
+        mcb = mcb.withCustomFilter(roleMapperConfig.getCustomLdapFilter());
+        ldapQuery.setModelCriteriaBuilder(mcb);
+
+        Stream<MapRoleEntity> ldapStream;
+
+        MapModelCriteriaBuilder<String,MapRoleEntity,RoleModel> mapMcb = queryParameters.getModelCriteriaBuilder().flashToModelCriteriaBuilder(createCriteriaBuilderMap());
+
+        Stream<LdapMapRoleEntityFieldDelegate> existingEntities = entities.entrySet().stream()
+                .filter(me -> mapMcb.getKeyFilter().test(keyConverter.fromString(me.getKey())) || deletedKeys.contains(me.getKey()))
+                .map(Map.Entry::getValue)
+                .filter(mapMcb.getEntityFilter())
+                // snapshot list
+                .collect(Collectors.toList()).stream();
+
+        try {
+            List<LdapMapObject> ldapObjects = identityStore.fetchQueryResults(ldapQuery);
+
+            ldapStream = ldapObjects.stream().map(ldapMapObject -> {
+                    // we might have fetch client and realm roles at the same time, now try to decode what is what
+                    StreamUtils.Pair<Boolean, String> client = getClientId(ldapMapObject.getDn());
+                    if (client == null) {
+                        return null;
+                    }
+                    LdapMapRoleEntityFieldDelegate entity = new LdapMapRoleEntityFieldDelegate(new LdapRoleEntity(ldapMapObject, roleMapperConfig, this, client.getV()));
+                    LdapMapRoleEntityFieldDelegate existingEntry = entities.get(entity.getId());
+                    if (existingEntry != null) {
+                        // this entry will be part of the existing entities
+                        return null;
+                    }
+                    entities.put(entity.getId(), entity);
+                    return (MapRoleEntity) entity;
+                })
+                .filter(Objects::nonNull)
+                // re-apply filters about client roles that we might have skipped for LDAP
+                .filter(me -> mapMcb.getKeyFilter().test(me.getId()))
+                .filter(me -> mapMcb.getEntityFilter().test(me))
+                // snapshot list
+                .collect(Collectors.toList()).stream();
```

Comment:
Yes, as discussed; I will update the comment.

Reply:
This is needed for the following scenario: the contents of the stream contents on `entities` and also updates `entities`, and two streams open at the same time could otherwise interfere.